### PR TITLE
🌱 Switch link checker to mdbook-linkcheck

### PR DESCRIPTION
**Pull request by benmoss**
**What this PR does / why we need it**:
Switches our link checker from https://github.com/raviqqe/liche to https://github.com/Michael-F-Bryan/mdbook-linkcheck/

This tool is directly integrated with mdbook which is nice, since it tests against rendered the rendered Markdown with our preprocessors rather than just scanning the raw files. It also supports adding headers for specific domains, which helps with avoiding rate-limiting by GitHub. It has built-in caching too, so links that have already been verified are faster.

I'm using my fork of https://github.com/japaric/trust/ which just gives us a nice succinct script to download Rust GitHub releases. We could also vendor the script. The upstream trust repo has a problem in that it requires `rustc` even if you're just using it to download a GitHub release, which I've made a PR to fix [here](https://github.com/japaric/trust/pull/132).

Removing `verify-book-links` as a part of `make verify` so that it doesn't become PR-blocking. I'm going to make a separate PR to test-infra to make it just a separate optional job, which for now should give us good enough visibility if a PR breaks a link or if a link breaks due to reasons outside our control (external servers changing).

Part of #2703

### CODE REVIEW

**Comment by benmoss** [comment]
We should add `GITHUB_TOKEN` to the Netlify environment variables to avoid hitting rate-limits https://docs.netlify.com/configure-builds/environment-variables/#deploy-request-notifications

File: docs/book/Makefile
Comment (vincepri) [inline docs/book/Makefile:42]: What's this doing? Seems rather something we shouldn't do here, curl and pipe to bash/sh usually raises security concerns. 

/hold

Comment (benmoss) [reply docs/book/Makefile:42]: Yes, I anticipated there might be concerns around this. We can vendor this script, but it's a script to install precompiled Rust crate published to GitHub. It's from https://github.com/japaric/trust but because of https://github.com/japaric/trust/pull/132 it's pulling from my fork.

Comment (benmoss) [reply docs/book/Makefile]: Updated to vendor the script, WDYT?

File: docs/book/Makefile
Comment (vincepri) [inline docs/book/Makefile:42]: Looks good, let's move the script maybe under `docs/third_party/crate_install.sh`? 

Other than that, 👍

File: docs/book/vendor/install.sh
Comment (vincepri) [inline docs/book/vendor/install.sh:2]: Is there a license for this file? Or was this created for this exact purpose?

Comment (benmoss) [reply docs/book/vendor/install.sh:2]: On the [master branch of the repo](https://github.com/japaric/trust) there's both an Apache and MIT license. Do you want that boilerplate-ified?

Comment (benmoss) [reply docs/book/vendor/install.sh]: They have a gh-pages branch that this file exists on normally, so it's served as a raw file.

Comment (vincepri) [reply docs/book/vendor/install.sh]: I'd defer to @detiber for license-y things 

I was mostly confused when opening the link, didn't find any license

File: docs/book/third_party/crate_install.sh
Comment (vincepri) [inline docs/book/third_party/crate_install.sh:2]: Let's have this script to be an unmodified copy of https://github.com/japaric/trust/blob/gh-pages/install.sh if possible, and include the LICENSE file https://github.com/japaric/trust/blob/master/LICENSE-APACHE

The directory structure for third party should be:
`/third_party/japaric-trust/...` (in the top level repo, where we already have other third party code

That should do it?

Comment (ncdc) [reply docs/book/third_party/crate_install.sh:2]: If it's unmodified, we'll need to put the target determination logic either in the makefile, or in a new script that determines the target, then calls crate_install.sh. If that's too much of a pain, we can keep things the way they are now, with the edits to the script.

Comment (benmoss) [reply docs/book/third_party/crate_install.sh:2]: Oh, true. I guess I assumed we wanted to leave the target unset so that developers on OSX and Linux would get the right binary, but we could wrap the script in another script that has the PR's logic to select the right target without rustc.

Comment (benmoss) [reply docs/book/third_party/crate_install.sh:2]: That all said, I think I prefer just modifying the script, it seems simpler.

**Comment by benmoss** [comment]
@vincepri any reason we keep pushing this one off? It's catching regressions we're missing

**Comment by vincepri** [comment]
Happy to merge whenever it's ready

**Comment by benmoss** [comment]
Should be ready now, I hadn't noticed the netlify failure

**Comment by benmoss** [comment]
@vincepri should be ready to go now, or at least review :smile:

**Review by vincepri — APPROVED** [review]
/approve
/hold cancel
/assign @ncdc @detiber

**Comment by ncdc** [comment]
Do we want the deploy/netlify preview to fail if there are broken links? It's not currently because the linkcheck is marked as optional and our netlify configuration doesn't install the link checker as part of the build.

**Comment by ncdc** [comment]
@detiber WDYT?

**Comment by detiber** [comment]
I'm a bit torn on failing the build...  In one sense it would be nice to get a heads up that things are broken and/or verify any new links added are not broken, but it also seems a bit harsh to penalize a contributor's PR if an existing link has become broken.

Unless there is a way to differentiate the two, then I think it should likely remain optional.

**Comment by ncdc** [comment]
/lgtm
😄 

That's fair. How do we anticipate getting signal that links are broken?

**Comment by detiber** [comment]
> That's fair. How do we anticipate getting signal that links are broken?

For netlify notifications, it looks like it is possible to configure both Slack and email notifications. 

For slack, it would require getting a slack token: https://github.com/kubernetes/community/blob/13b3f6a596f3b31c668409c1a52f386f98eab93f/.github/ISSUE_TEMPLATE/slack-request.md

**Pull request merged**
